package racingcar;

import camp.nextstep.edu.missionutils.test.NsTest;
import org.junit.jupiter.api.Test;

import static camp.nextstep.edu.missionutils.test.Assertions.assertRandomNumberInRangeTest;
import static camp.nextstep.edu.missionutils.test.Assertions.assertSimpleTest;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class ApplicationTest extends NsTest {
    private static final int MOVING_FORWARD = 4;
    private static final int STOP = 3;

    @Test
    void 전진_정지() {
        assertRandomNumberInRangeTest(
            () -> {
                run("pobi,woni", "1");
                assertThat(output()).contains("pobi : -", "woni : ", "최종 우승자 : pobi");
            },
            MOVING_FORWARD, STOP
        );
    }

    @Test
    void 이름이_5글자_이상일_시_예외_처리() {
        assertSimpleTest(() ->
                assertThatThrownBy(() -> runException("pobi,javaji", "1"))
                        .isInstanceOf(IllegalArgumentException.class)
        );
    }

    @Test
    void 이름이_1글자_미만일_시_예외_처리() {
        assertSimpleTest(() ->
                assertThatThrownBy(() -> runException("one,,three", "1"))
                        .isInstanceOf(IllegalArgumentException.class)
        );
    }

    @Test
    void 이름_입력값에_공백이_포함될_시_예외_처리() {
        assertSimpleTest(() ->
                assertThatThrownBy(() -> runException("one, two ,three", "1"))
                        .isInstanceOf(IllegalArgumentException.class)
        );
    }

    @Test
    void 이름이_중복으로_입력될_시_예외_처리() {
        assertSimpleTest(() ->
                assertThatThrownBy(() -> runException("one,one", "1"))
                        .isInstanceOf(IllegalArgumentException.class)
        );
    }

    @Test
    void 시행_횟수_입력값이_정수형이_아닐_경우_예외_처리() {
        assertSimpleTest(() ->
                assertThatThrownBy(() -> runException("one,two,three", "abc"))
                        .isInstanceOf(IllegalArgumentException.class)
        );
    }

    @Test
    void 시행_횟수_입력값이_0일_경우_예외_처리() {
        assertSimpleTest(() ->
                assertThatThrownBy(() -> runException("one,two,three", "0"))
                        .isInstanceOf(IllegalArgumentException.class)
        );
    }

    @Test
    void 시행_횟수_입력값이_음수일_경우_예외_처리() {
        assertSimpleTest(() ->
                assertThatThrownBy(() -> runException("one,two,three", "-1"))
                        .isInstanceOf(IllegalArgumentException.class)
        );
    }

    @Override
    public void runMain() {
        Application.main(new String[]{});
    }
}
